firedrake/matrix_free/preconditioners.py

```diff
@@ -0,0 +1,331 @@
+    velocity.  You must provide these with options using the glbaol
+    option ``Re`` for the Reynolds number and the prefixed option
+    ``pcd_velocity_space`` which should be the index into the full
+    space that gives velocity field.
+
+    """
+    def initialize(self, pc):
+        from firedrake import TrialFunction, TestFunction, dx, \
+            assemble, inner, grad, split, Constant
+        prefix = pc.getOptionsPrefix() + "pcd_"
+
+        # we assume P has things stuffed inside of it
+        _, P = pc.getOperators()
+        context = P.getPythonContext()
+
+        test, trial = context.a.arguments()
+        if test.function_space() != trial.function_space():
+            raise ValueError("Pressure space test and trial space differ")
+
+        Q = test.function_space()
+
+        p = TrialFunction(Q)
+        q = TestFunction(Q)
+
+        mass = p*q*dx
+
+        # Regularisation to avoid having to think about nullspaces.
+        stiffness = inner(grad(p), grad(q))*dx + Constant(1e-6)*p*q*dx
+
+        opts = PETSc.Options()
+        # we're inverting Mp and Kp, so default them to assembled.
+        # Fp only needs its action, so default it to mat-free.
+        # These can of course be overridden.
+        # only Fp is referred to in update, so that's the only
+        # one we stash.
+        Mp_mat_type = opts.getString(prefix+"Mp_mat_type", "")
+        Kp_mat_type = opts.getString(prefix+"Kp_mat_type", "")
+        self.Fp_mat_type = opts.getString(prefix+"Fp_mat_type", "matfree")
+
+        
+        # FIXME: allow these guys to be matrix-free
+        Mp = assemble(mass, form_compiler_parameters=context.fc_params,
+                      mat_type=Mp_mat_type)
+        Kp = assemble(stiffness, form_compiler_parameters=context.fc_params,
+                      mat_type=Kp_mat_type)
+
+        Mp.force_evaluation()
+        Kp.force_evaluation()
+
+        # FIXME: Should we transfer nullspaces over.  I think not.
+
+        Mksp = PETSc.KSP().create()
+        Mksp.setOptionsPrefix(prefix + "Mp_")
+        Mksp.setOperators(Mp.petscmat)
+        Mksp.setUp()
+        Mksp.setFromOptions()
+        self.Mksp = Mksp
+
+        Kksp = PETSc.KSP().create()
+        Kksp.setOptionsPrefix(prefix + "Kp_")
+        Kksp.setOperators(Kp.petscmat)
+        Kksp.setUp()
+        Kksp.setFromOptions()
+        self.Kksp = Kksp
+
+        state = context.context["state"]
+
+        Re = context.context.get("Re", 1.0)
+
+        velid = context.context["velocity_space"]
+
+        u0 = split(state)[velid]
+        fp = 1.0/Re * inner(grad(p), grad(q))*dx + inner(u0, grad(p))*q*dx
+
+        self.Re = Re
+        # FIXME, allow assembled matrix here
+        self.Fp = assemble(fp, form_compiler_parameters=context.fc_params,
+                           mat_type=self.Fp_mat_type)
+        self.Fp.force_evaluation()
+        Fpmat = self.Fp.petscmat
+        self.workspace = [Fpmat.createVecLeft() for i in (0, 1)]
```

Comment:
I'd like to have it present for users, even if in a limited case and with
documentation.  But maybe we can fix this.  I picked the right bc's for
characteristic boundary, which
worked for driven cavity and for enclosed flow like Rayleigh-Benard.  But I
knew it would need extension for, e.g. backward step.

The question is how to identify "the part of the boundary where u_0.n is
positive" and a.) specify DirichletBC on that or b.) use it in a measure ds
(needed for Robin, Neumann, and weak Dirichlet BC's)

If there is a clean way to get this, there is no reason not to extend the
PCDPC to DTRT.  It could get a bit fussy if we want to allow both strong
and Nitsche-type BC's.


On Wed, Aug 17, 2016 at 4:11 AM, Lawrence Mitchell <notifications@github.com
> wrote:

> In firedrake/matrix_free/preconditioners.py
> <https://github.com/firedrakeproject/firedrake/pull/853#discussion_r75087747>
> :
>
> > +        state = context.context["state"]
> > +
> > +        Re = context.context.get("Re", 1.0)
> > +
> > +        velid = context.context["velocity_space"]
> > +
> > +        u0 = split(state)[velid]
> > +        fp = 1.0/Re * inner(grad(p), grad(q))*dx + inner(u0, grad(p))*q*dx
> > +
> > +        self.Re = Re
> > +        # FIXME, allow assembled matrix here
> > +        self.Fp = assemble(fp, form_compiler_parameters=context.fc_params,
> > +                           mat_type=self.Fp_mat_type)
> > +        self.Fp.force_evaluation()
> > +        Fpmat = self.Fp.petscmat
> > +        self.workspace = [Fpmat.createVecLeft() for i in (0, 1)]
>
> @pefarrell <https://github.com/pefarrell> reminds me that the treatment
> of boundary conditions in PCD is somewhat subtle. I think the best solution
> is in https://www.cs.umd.edu/~elman/papers/bc_approx_comm.pdf. Given
> this, it suggests that we don't want PCD to be a builtin preconditioner
> (since handling all the correct inflow/outflow boundaries is rather problem
> dependent), e.g. dirichlet inflow conditions turn up as robin conditions on
> Fp.
>
> Maybe we should instead incorporate this into the demos as an example
> "user-defined" preconditioner.
>
> What do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/firedrakeproject/firedrake/pull/853/files/4ae09350675779089c0b84e43dd28c13a7d4940f#r75087747>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADl6L2TU0l-B-kFODELQph2_8inBGLKVks5qgtAmgaJpZM4JiV-g>
> .
>